### ISSUE TITLE
Converting decimal loses precision

### ISSUE THREAD

**Issue by boumboum** [mkleehammer/pyodbc#333]
I have this insert:
sql = "insert into mytable (id, km, dij_km) values(?, ?, ?)"
cr.execute(sql, 6624, Decimal('2.503'), Decimal('0.0999999999999996'))
I have an error message:
(<class 'pyodbc.ProgrammingError'>, ProgrammingError('Converting decimal loses precision', 'HY000')
table definition:
create table bd_mytable 
	(
	id int constraint PK_bd_vonal_pont PRIMARY KEY NONCLUSTERED,
	km decimal(7,3) NOT NULL,
	dij_km decimal(7,3) default(0)
	)

If I do the insert within MSSM, there is no error at all.
It seems pyodbc doesn't allow the implicit rounding. Why?

- Python: 3.6
- pyodbc: 4022
- OS: Win7
- DB: MSSQL 2014
- driver: ODBC13

**Comment by v-chojas** [comment]
This is by design. Decimal is an exact datatype so losing precision will cause an error. If you want to e.g.  truncate a Decimal(16,16) to Decimal(7,3), then you should do that explicitly.

See https://github.com/mkleehammer/pyodbc/issues/302 for more information.

**Comment by boumboum** [comment]
Yes, but in this case the design is bad.
I have different kind of precisions in the database so when I import the data (from a big xml file) I cannot just round to a predefine precision because a priori I do not know the precision defined within the database.
And it must be of the same type, because I'm doing a comparison and a float(1.2) is not equal to Decimal(1.2)

**Comment by v-chojas** [comment]
Either change the database columns to match the precision of the data you are inserting, or change the data to match. The goal of the design is to prevent any silent data loss from occurring.

**Comment by boumboum** [comment]
I understand what you are saying but I think a corresponding rounding under the hood would do the job.
I cannot change the DB design, as it doesn't belong to my authority and my problem is exactly the data match.
I is basically a string as I'm consuming an XML file. So I have to convert the data to the corresponding format.
And as I've told you in my first post, MSSQL doesn't do any problem getting different precision value, just do the rounding silently.
i.e if you have a table 
```
create table #t(a decimal(5,2))
insert into #t values(10.145)
select * from #t
```
the result is 10.15
Without error, without warning.
Basically this is what I'm expecting from pyodbc as well.
(btw: without wanting to do any comparison between the two, but pymssql do the insert without issue. So it is doable :-( )

**Comment by v-chojas** [comment]
If it is a string then you should be able to insert them as such, letting the server do the conversion.
In any case I was unable to reproduce your error with latest pyodbc; inserting Decimal('0.0999999999999996') into a decimal(7,3) column works and becomes a Decimal('0.100') when retrieved.

**Comment by mkleehammer** [comment]
@v-chojas The error only occurs when the value is being bound as a numeric structure.

Since the database itself does not have this behavior, I'm inclined to agree with the reporter.  This is something pyodbc added.  If it was added to make the implementation more convenient, we definitely need to re-evaluate it.  If it is for safety, I'd let the DB handle it.  If there is a deeper reason, we should see if there is a way around it.

**Comment by boumboum** [comment]
Exact. If I want to insert a python type decimal.Decimal with a "wrong" precision, the insert fails.
If I do the insert with a character representation of the value, there is no error.
I also tend to say that pyodbc should leave the validation to the DB server.

**Comment by v-chojas** [comment]
@mkleehammer Binding decimal as decimal/numeric was done to avoid unnecessary conversion to and from strings (performance, possible precision issues too). If this breaks existing behaviour then I agree about re-evaluating it. Perhaps it should be made optional.

**Comment by mkleehammer** [comment]
I'm closing this because I am pretty sure that the parameter reworking (and undoing) fixed this, at least in the normal path.
